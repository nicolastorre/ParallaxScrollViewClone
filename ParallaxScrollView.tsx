import React, {useRef} from 'react';
import {
  Animated,
  Platform,
  StatusBar,
  StyleSheet,
  Text,
  View,
} from 'react-native';

type ParallaxScrollViewProps = {
  title?: string | React.ReactNode;
  renderContent: () => React.ReactNode;
  headerMaxHeight?: number;
  headerMinHeight?: number;
};

const STATUS_BAR_HEIGHT =
  Platform.OS === 'ios' ? 44 : StatusBar.currentHeight ?? 0;

export const ParallaxScrollView: React.FC<ParallaxScrollViewProps> = ({
  title,
  renderContent,
  headerMaxHeight = 170,
  headerMinHeight = 64,
}) => {
  const scrollY = useRef(new Animated.Value(0)).current;

  const getInputRange = () => [0, headerMaxHeight - headerMinHeight];

  const headerHeight = scrollY.interpolate({
    inputRange: getInputRange(),
    outputRange: [headerMaxHeight, headerMinHeight],
    extrapolate: 'clamp',
  });

  const scrollEventThrottle = 16;

  return (
    <View style={styles.container}>
      <Animated.ScrollView
        style={styles.scrollView}
        scrollEventThrottle={scrollEventThrottle}
        onScroll={Animated.event(
          [{nativeEvent: {contentOffset: {y: scrollY}}}],
          {
            useNativeDriver: false,
          },
        )}>
        <View style={{marginTop: headerMaxHeight}}>{renderContent()}</View>
      </Animated.ScrollView>

      <Animated.View
        style={[
          styles.header,
          {
            height: headerHeight,
          },
        ]}>
        {typeof title === 'string' ? (
          <Text style={styles.headerText}>{title}</Text>
        ) : (
          title
        )}
      </Animated.View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  scrollView: {
    flex: 1,
  },
  header: {
    backgroundColor: 'black',
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    alignItems: 'center',
    justifyContent: 'center',
    paddingTop: STATUS_BAR_HEIGHT,
  },
  headerText: {
    color: 'white',
    textAlign: 'center',
    fontSize: 16,
  },
});
