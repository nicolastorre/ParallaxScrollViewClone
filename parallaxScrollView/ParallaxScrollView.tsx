import React, {useMemo, useRef} from 'react';
import {Animated, StyleSheet, View} from 'react-native';
import {common} from '../src/styles/common';

type ParallaxScrollViewProps = {
  mainHeader: React.ReactNode;
  navHeader: React.ReactNode;
  renderContent: () => React.ReactNode;
  headerMaxHeight?: number;
  headerMinHeight?: number;
  scrollEventThrottle?: number;
  headerBackgroundColor?: string;
  fadeDistance?: number;
};

export const ParallaxScrollView: React.FC<ParallaxScrollViewProps> = ({
  mainHeader,
  navHeader,
  renderContent,
  headerMaxHeight = 90,
  headerMinHeight = 40,
  scrollEventThrottle = 16,
  headerBackgroundColor = 'black',
  fadeDistance = 0,
}) => {
  const scrollY = useRef(new Animated.Value(0)).current;

  const scrollRange = useMemo(
    () => headerMaxHeight - headerMinHeight,
    [headerMaxHeight, headerMinHeight],
  );

  const opacityMainHeader = useMemo(
    () =>
      scrollY.interpolate({
        inputRange: [0, scrollRange - fadeDistance],
        outputRange: [1, 0],
        extrapolate: 'clamp',
      }),
    [scrollY, scrollRange, fadeDistance],
  );

  const opacityNavHeader = useMemo(
    () =>
      scrollY.interpolate({
        inputRange: [scrollRange - fadeDistance, scrollRange],
        outputRange: [0, 1],
        extrapolate: 'clamp',
      }),
    [scrollY, scrollRange, fadeDistance],
  );

  const navHeaderStyle = useMemo(
    () => [
      styles.headerContainer,
      {
        backgroundColor: headerBackgroundColor,
        opacity: opacityNavHeader,
        zIndex: 1000,
      },
    ],
    [headerBackgroundColor, opacityNavHeader],
  );

  return (
    <View style={common.container}>
      <Animated.View
        style={[
          styles.headerContainer,
          {
            backgroundColor: headerBackgroundColor,
            opacity: opacityMainHeader,
          },
        ]}
        accessible
        accessibilityLabel="Main Header"
        testID="main-header">
        {mainHeader}
      </Animated.View>

      <Animated.View
        style={navHeaderStyle}
        accessible
        accessibilityLabel="Navigation Header"
        testID="nav-header">
        {navHeader}
      </Animated.View>

      <Animated.ScrollView
        style={styles.scrollView}
        scrollEventThrottle={scrollEventThrottle}
        contentContainerStyle={{paddingTop: headerMaxHeight}}
        onScroll={Animated.event(
          [{nativeEvent: {contentOffset: {y: scrollY}}}],
          {
            useNativeDriver: false,
          },
        )}>
        {renderContent()}
      </Animated.ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  scrollView: {
    flex: 1,
  },
  headerContainer: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
