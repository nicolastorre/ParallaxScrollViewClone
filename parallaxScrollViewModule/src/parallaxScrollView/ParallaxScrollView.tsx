import React, {useRef} from 'react';
import {Animated, StyleSheet, View} from 'react-native';
import {AnimatedHeader} from '../animatedHeader/AnimatedHeader';
import {common} from '../../../src/styles/common';

type ParallaxScrollViewProps = {
  mainHeader: React.ReactNode;
  navHeader: React.ReactNode;
  renderContent: () => React.ReactNode;
  headerMaxHeight?: number;
  headerMinHeight?: number;
  scrollEventThrottle?: number;
  headerBackgroundColor?: string;
  fadeDistance?: number;
};

export const ParallaxScrollView: React.FC<ParallaxScrollViewProps> = ({
  mainHeader,
  navHeader,
  renderContent,
  headerMaxHeight = 80,
  headerMinHeight = 40,
  scrollEventThrottle = 16,
  headerBackgroundColor = 'black',
  fadeDistance = 0,
}) => {
  const scrollY = useRef(new Animated.Value(0)).current;

  return (
    <View style={common.container}>
      <AnimatedHeader
        scrollY={scrollY}
        headerMaxHeight={headerMaxHeight}
        headerMinHeight={headerMinHeight}
        mainHeader={mainHeader}
        navHeader={navHeader}
        headerBackgroundColor={headerBackgroundColor}
        fadeDistance={fadeDistance}
      />

      <Animated.ScrollView
        style={styles.scrollView}
        scrollEventThrottle={scrollEventThrottle}
        contentContainerStyle={{paddingTop: headerMaxHeight}}
        onScroll={Animated.event(
          [{nativeEvent: {contentOffset: {y: scrollY}}}],
          {
            useNativeDriver: false,
          },
        )}>
        {renderContent()}
      </Animated.ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  scrollView: {
    flex: 1,
  },
});
