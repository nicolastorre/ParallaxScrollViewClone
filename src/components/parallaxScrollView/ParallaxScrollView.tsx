import React, {useRef} from 'react';
import {Animated, StyleSheet, View} from 'react-native';
import {AnimatedHeader} from '../animatedHeader/AnimatedHeader';
import {common} from '../../styles/common';

type ParallaxScrollViewProps = {
  mainHeader: React.ReactNode;
  navHeader: React.ReactNode;
  renderContent: () => React.ReactNode;
  headerMaxHeight?: number;
  headerMinHeight?: number;
  scrollEventThrottle?: number;
  headerBackgroundColor?: string;
};

export const ParallaxScrollView: React.FC<ParallaxScrollViewProps> = ({
  mainHeader,
  navHeader,
  renderContent,
  headerMaxHeight = 170,
  headerMinHeight = 64,
  scrollEventThrottle = 16,
  headerBackgroundColor = 'black',
}) => {
  const scrollY = useRef(new Animated.Value(0)).current;

  return (
    <View style={common.container}>
      <AnimatedHeader
        scrollY={scrollY}
        headerMaxHeight={headerMaxHeight}
        headerMinHeight={headerMinHeight}
        mainHeader={mainHeader}
        navHeader={navHeader}
        headerBackgroundColor={headerBackgroundColor}
      />

      <Animated.ScrollView
        style={styles.scrollView}
        scrollEventThrottle={scrollEventThrottle}
        contentContainerStyle={{paddingTop: headerMaxHeight}}
        onScroll={Animated.event(
          [{nativeEvent: {contentOffset: {y: scrollY}}}],
          {
            useNativeDriver: false,
          },
        )}>
        {renderContent()}
      </Animated.ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  scrollView: {
    flex: 1,
  },
});
