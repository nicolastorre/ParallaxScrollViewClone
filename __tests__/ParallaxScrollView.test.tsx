import React from 'react';
import {render} from '@testing-library/react-native';
import {ParallaxScrollView} from '../ParallaxScrollView';
import {Text} from 'react-native';

const mainHeader = 'Test main header';
const navHeader = 'Test nav header';
const contentText = 'Test Content Item';

describe('ParallaxScrollView', () => {
  it('renders the title and content correctly', () => {
    const {getByText} = render(
      <ParallaxScrollView
        mainHeader={mainHeader}
        navHeader={navHeader}
        renderContent={() => (
          <>
            <Text>{contentText}</Text>
          </>
        )}
      />,
    );

    expect(getByText(mainHeader)).toBeTruthy();

    expect(getByText(contentText)).toBeTruthy();
  });

  it('matches snapshot', () => {
    const tree = render(
      <ParallaxScrollView
        mainHeader={mainHeader}
        navHeader={navHeader}
        renderContent={() => (
          <>
            <Text>Item 1</Text>
            <Text>Item 2</Text>
          </>
        )}
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
