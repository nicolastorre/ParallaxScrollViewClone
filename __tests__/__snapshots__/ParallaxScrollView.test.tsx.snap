// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParallaxScrollView matches snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    collapsable={false}
    contentContainerStyle={
      {
        "paddingTop": 170,
      }
    }
    onScroll={[Function]}
    scrollEventThrottle={16}
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <Text>
        Item 1
      </Text>
      <Text>
        Item 2
      </Text>
    </View>
  </RCTScrollView>
  <View
    accessibilityLabel="Parallax Header Container"
    accessible={true}
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "black",
        "height": 170,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 1000,
      }
    }
    testID="parallax-header"
  >
    <RCTSafeAreaView>
      <View
        accessibilityLabel="Main Header"
        accessible={true}
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
            "opacity": 1,
          }
        }
        testID="main-header"
      >
        <Text
          accessibilityLabel="Main header title: Test main header"
          accessibilityRole="header"
          style={
            {
              "color": "white",
              "fontSize": 16,
              "textAlign": "center",
            }
          }
          testID="header-title"
        >
          Test main header
        </Text>
      </View>
      <View
        accessibilityLabel="Navigation Header"
        accessible={true}
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
            "opacity": 0,
          }
        }
        testID="nav-header"
      >
        <Text
          accessibilityLabel="Nav header title: Test nav header"
          accessibilityRole="header"
          style={
            {
              "color": "white",
              "fontSize": 16,
              "textAlign": "center",
            }
          }
          testID="nav-header-title"
        >
          Test nav header
        </Text>
      </View>
    </RCTSafeAreaView>
  </View>
</View>
`;
